File: csharp/main/LiteralInt8.cs
using System;

namespace Deltix.Luminary
{
	public sealed class LiteralInt8 : LiteralInteger, ILiteral<SByte>
	{
		public LiteralInt8(SByte value)
			: base(LiteralKind.SInt8)
		{
			Value = value;
		}

		public SByte Value { get; }

		public override LiteralInteger CastTo(IntegralType type)
		{
			switch (type)
			{
				case IntegralType.Int64:
					return new LiteralInt64(Value);

				case IntegralType.UInt64:
					if (Value < 0)
						throw new InvalidCastException($"Value '{Value}' cannot be safely cast to 'UInt64'.");
					return new LiteralUInt64((UInt64) Value);

				case IntegralType.Int32:
					return new LiteralInt32(Value);

				case IntegralType.UInt32:
					if (Value < 0)
						throw new InvalidCastException($"Value '{Value}' cannot be safely cast to 'UInt32'.");
					return new LiteralUInt32((UInt32) Value);

				case IntegralType.Int16:
					return new LiteralInt64(Value);

				case IntegralType.UInt16:
					if (Value < 0)
						throw new InvalidCastException($"Value '{Value}' cannot be safely cast to 'UInt16'.");
					return new LiteralUInt16((UInt16) Value);

				case IntegralType.Int8:
					return this;

				case IntegralType.UInt8:
					if (Value < 0)
						throw new InvalidCastException($"Value '{Value}' cannot be safely cast to 'UInt8'.");
					return new LiteralUInt8((Byte) Value);

				default:
					throw new ArgumentOutOfRangeException(nameof(type), type, null);
			}
		}

		public override String ToString()
		{
			return $"{Value}i8";
		}
	}
}

File: java/src/main/java/deltix/luminary/LiteralInt8.java
package deltix.luminary;

public class LiteralInt8 extends LiteralInteger {
    private final byte value;

    public LiteralInt8(byte value) {
        super(LiteralKind.INT8);
        this.value = value;
    }

    public byte getValue() {
        return value;
    }

    @Override
    public LiteralInteger castTo(IntegralType type) {
        switch (type) {
            case INT8:
                return this;

            case UINT8:
                if (value < 0)
                    throw new ClassCastException("Value '" + value + "' cannot be safely cast to 'UInt8'.");
                return new LiteralUInt8(value);

            case INT16:
                return new LiteralInt16(value);

            case UINT16:
                if (value < 0)
                    throw new ClassCastException("Value '" + value + "' cannot be safely cast to 'UInt16'.");
                return new LiteralUInt16(value);

            case INT32:
                return new LiteralInt32(value);

            case UINT32:
                if (value < 0)
                    throw new ClassCastException("Value '" + value + "' cannot be safely cast to 'UInt32'.");
                return new LiteralUInt32(value);

            case INT64:
                return new LiteralInt64(value);

            case UINT64:
                if (value < 0)
                    throw new ClassCastException("Value '" + value + "' cannot be safely cast to 'UInt64'.");
                return new LiteralUInt64(value);

            default:
                throw new IllegalArgumentException("Undefined integral type.");
        }
    }

    @Override
    public String toString() {
        return value + "i8";
    }
}
